### PR TITLE
[SPARK-33354][SQL] New explicit cast syntax rules in ANSI mode

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

In section 6.13 of the ANSI SQL standard, there are syntax rules for valid combinations of the source and target data types.
image

Comparing the ANSI CAST syntax rules with the current default behavior of Spark:
image

To make Spark's ANSI mode more ANSI SQL Compatible，I propose to disallow the following casting in ANSI mode:
```
TimeStamp <=> Boolean
Date <=> Boolean
Numeric <=> Timestamp
Numeric <=> Date
Numeric <=> Binary
String <=> Array
String <=> Map
String <=> Struct
```
The following castings are considered invalid in ANSI SQL standard, but they are quite straight forward. Let's Allow them for now
```
Numeric <=> Boolean
String <=> Binary
```
### Why are the changes needed?

Better ANSI SQL compliance

### Does this PR introduce _any_ user-facing change?

Yes, the following castings will not be allowed in ANSI mode:
```
TimeStamp <=> Boolean
Date <=> Boolean
Numeric <=> Timestamp
Numeric <=> Date
Numeric <=> Binary
String <=> Array
String <=> Map
String <=> Struct
```

### How was this patch tested?

Unit test

The ANSI Compliance doc preview:
image

### CODE REVIEW

**Comment by gengliangwang** [comment]
I am still adding more test case, mark it as WIP for now

**Comment by maropu** [comment]
Thanks for the request, @gengliangwang ! I'll review this later.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 1]
```diff
@@ -1770,6 +1782,64 @@ case class AnsiCast(child: Expression, dataType: DataType, timeZoneId: Option[St
     copy(timeZoneId = Option(timeZoneId))
 
   override protected val ansiEnabled: Boolean = true
+
+  override def canCast(from: DataType, to: DataType): Boolean = AnsiCast.canCast(from, to)
+}
+
+object AnsiCast {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:1789]: Could you leave some comments to summarize the current behaivour of the ANSI explicit cast as described in the PR description (references `section 6.13 of the ANSI SQL standard` and differences from the standard, e.g., `Numeric <=> Boolean`) ?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 2]
```diff
@@ -38,8 +38,11 @@ import org.apache.spark.unsafe.types.UTF8String
 
 abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
 
+  // Whether this test suite is for checking AnsiCast or default Cast.
+  protected def isAnsiCast: Boolean
+
   // Whether it is required to set SQLConf.ANSI_ENABLED as true for testing numeric overflow.
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:44]: I feel the name `requiredAnsiEnabledForOverflowTestCases` is confusing, so could you rename it according to the current usage and update the comment above?

**Review by maropu — COMMENTED** [review]
Also, I think we need to update the migration guide, too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 3]
```diff
@@ -1770,6 +1782,64 @@ case class AnsiCast(child: Expression, dataType: DataType, timeZoneId: Option[St
...
+  def canCast(from: DataType, to: DataType): Boolean = (from, to) match {
+    case (fromType, toType) if fromType == toType => true
+
+    case (NullType, _) => true
+
+    case (_: NumericType, _: NumericType) => true
+    case (StringType, _: NumericType) => true
+    case (BooleanType, _: NumericType) => true
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:1797]: (Just a suggestion) For readability, could you reorder these entries according to `Cast.canCast` where possible? For example, the numeric entries for `Cast` seems to be placed just before complex types:
https://github.com/apache/spark/blob/35ac314181129374b02f8f8c07341b43a734e1c7/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L70-L74

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 4]
```diff
@@ -1396,14 +1337,125 @@ class CastSuite extends CastSuiteBase {
 
     checkEvaluation(cast("abcd", DecimalType(38, 1)), null)
   }
+
+  test("SPARK-22825 Cast array to string") {
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:1341]: (nit and this is not related to this PR though...) `SPARK-22825` -> `SPARK-22825:`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 5]
```diff
@@ -1396,14 +1337,125 @@ class CastSuite extends CastSuiteBase {
...
+      withSQLConf(SQLConf.LEGACY_COMPLEX_TYPES_TO_STRING.key -> omitNull.toString) {
+        val ret1 = cast(Literal.create(Array(null, null)), StringType)
+        checkEvaluation(
+          ret1,
+          s"[${if (omitNull) "" else "null"},${if (omitNull) "" else " null"}]")
+      }
+    }
+  }
+
+  test("SPARK-22973 Cast map to string") {
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:1383]: ditto

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 6]
```diff
@@ -1396,14 +1337,125 @@ class CastSuite extends CastSuiteBase {
...
+          Literal.create(Map(
+            1 -> Array(1, 2, 3),
+            2 -> Array(4, 5, 6))),
+          StringType)
+        checkEvaluation(ret5, s"${lb}1 -> [1, 2, 3], 2 -> [4, 5, 6]$rb")
+      }
+    }
+  }
+
+  test("SPARK-22981 Cast struct to string") {
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:1417]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala [hunk 7]
```diff
@@ -756,6 +756,47 @@ class InsertSuite extends DataSourceTest with SharedSparkSession {
     }
   }
 
+  test("SPARK-33354: Throw exceptions on inserting invalid cast with ANSI casting policy") {
+    withSQLConf(
+      SQLConf.STORE_ASSIGNMENT_POLICY.key -> SQLConf.StoreAssignmentPolicy.ANSI.toString) {
+      withTable("t") {
+        sql("create table t(i int, t timestamp) using parquet")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/sources/InsertSuite.scala:763]: nit: please use uppercases for SQL keywords, e.g., `create` -> `CREATE`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 8]
```diff
@@ -1753,6 +1759,12 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
     copy(timeZoneId = Option(timeZoneId))
 
   override protected val ansiEnabled: Boolean = SQLConf.get.ansiEnabled
+
+  override def canCast(from: DataType, to: DataType): Boolean = if (ansiEnabled) {
```
[the remark this answers is not in the document]
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:1763]: well, then we need to mention about the behavior of throwing overflow exceptions when ANSI flag enabled. I will add some content in the `sql-ref-ansi-compliance.md`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 4 above)
Comment (gengliangwang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:1341]: I would prefer keeping them unchanged in this PR

**Comment by gengliangwang** [comment]
@maropu thanks for the suggestions. I have updated the code. Please take another look.

File: docs/sql-ref-ansi-compliance.md [hunk 9]
```diff
@@ -107,6 +107,22 @@ SELECT * FROM t;
 +---+
 ```
 
+The valid combinations of target data type and source data type in a `Cast` expression are given by the following table.
```
Comment (maropu) [inline docs/sql-ref-ansi-compliance.md:110]: How about moving this statements into L61-62?

Comment (maropu) [reply docs/sql-ref-ansi-compliance.md:110]: nit: `Cast` expression -> `CAST` syntax ?

File: docs/sql-ref-ansi-compliance.md [hunk 10]
```diff
@@ -107,6 +107,22 @@ SELECT * FROM t;
...
+| Numeric   | Y       | Y      | N    | N         | N        | Y       | N      | N     | N   | N      |
+| String    | Y       | Y      | Y    | Y         | Y        | Y       | Y      | N     | N   | N      |
+| Date      | N       | Y      | Y    | Y         | N        | N       | N      | N     | N   | N      |
+| Timestamp | N       | Y      | Y    | Y         | N        | N       | N      | N     | N   | N      |
+| Interval  | N       | Y      | N    | N         | Y        | N       | N      | N     | N   | N      |
+| Boolean   | Y       | Y      | N    | N         | N        | Y       | N      | N     | N   | N      |
+| Binary    | Y       | N      | N    | N         | N        | N       | Y      | N     | N   | N      |
+| Array     | N       | N      | N    | N         | N        | N       | N      | Y     | N   | N      |
+| Map       | N       | N      | N    | N         | N        | N       | N      | N     | Y   | N      |
+| Struct    | N       | N      | N    | N         | N        | N       | N      | N     | N   | Y      |
```
Comment (maropu) [inline docs/sql-ref-ansi-compliance.md:124]: How about following the wording in the data type document, e.g., `String` -> `StringType` https://spark.apache.org/docs/3.0.1/sql-ref-datatypes.html

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 11]
```diff
@@ -1770,6 +1782,110 @@ case class AnsiCast(child: Expression, dataType: DataType, timeZoneId: Option[St
     copy(timeZoneId = Option(timeZoneId))
 
   override protected val ansiEnabled: Boolean = true
+
+  override def canCast(from: DataType, to: DataType): Boolean = AnsiCast.canCast(from, to)
+}
+
+object AnsiCast {
+  /**
+   * As per section 6.13 "cast specification" in "Information technology — Database languages " +
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:1791]: Nice update. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 12]
```diff
@@ -1770,6 +1782,110 @@ case class AnsiCast(child: Expression, dataType: DataType, timeZoneId: Option[St
...
+   *   UDT  = User-Defined Type
+   *   B   = Binary (Fixed- or Variable-Length or Binary Large Object)
+   *   RT  = Reference type
+   *   CT  = Collection type
+   *   RW  = Row type
+   *
+   * Spark's ANSI mode follows the syntax rules, except it specially allow the following
+   * straightforward type conversions which are disallowed as per the SQL standard:
+   *   - Numeric <=> Boolean
+   *   - String <=> Binary
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:1834]: How about describing this in `sql-ref-ansi-compliance.md`, too?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 13]
```diff
@@ -635,16 +585,74 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
...
-    assert(cast(1, DateType).checkInputDataTypes().isFailure)
-    assert(cast(1L, DateType).checkInputDataTypes().isFailure)
-    assert(cast(1.0.toFloat, DateType).checkInputDataTypes().isFailure)
-    assert(cast(1.0, DateType).checkInputDataTypes().isFailure)
+    checkInvalidCastFromNumericType(DateType)
+  }
+
+  test("ANSI mode: disallow type conversions between Numeric types and Timestamp type") {
+    import DataTypeTestUtils.numericTypes
+    withSQLConf(ansiConfForExceptionOnOverflow) {
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:606]: Is this test related to overflow tests? (Seems the config name suggests so)

**Review by maropu — COMMENTED** [review]
Looks fine otherwise. cc: @cloud-fan

**Comment by gengliangwang** [comment]
@maropu Thanks again for the thoughtful suggestions! I have updated the code.

**Comment by maropu** [comment]
kindly ping @cloud-fan

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 14]
```diff
@@ -930,150 +733,254 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
...
+class CastSuite extends CastSuiteBase {
+
+  override def cast(v: Any, targetType: DataType, timeZoneId: Option[String] = None): CastBase = {
+    v match {
+      case lit: Expression => Cast(lit, targetType, timeZoneId)
+      case _ => Cast(Literal(v), targetType, timeZoneId)
+    }
+  }
+
+  test("null cast II") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:929]: nit: `null cast #2`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 15]
```diff
@@ -1396,93 +1355,199 @@ class CastSuite extends CastSuiteBase {
...
-      checkExceptionInExpression[NumberFormatException](
-        cast("1.23", dataType), "invalid input syntax for type numeric: 1.23")
-    }
+/**
+ * Test suite for data type casting expression [[AnsiCast]] with ANSI mode enabled.
+ */
+class AnsiCastSuiteWithAnsiModeOn extends CastSuiteBase {
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    SQLConf.get.setConf(SQLConf.ANSI_ENABLED, true)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:1517]: It's fragile to change global conf. How about we override `test` method and wrap the test funnc with `withSQLConf`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 16]
```diff
@@ -930,150 +733,254 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
...
+    numericTypes.foreach(dt => checkNullCast(TimestampType, dt))
+    numericTypes.foreach(dt => checkNullCast(DateType, dt))
+  }
+
+  test("cast from int") {
+    checkCast(0, false)
+    checkCast(1, true)
+    checkCast(-5, true)
+    checkCast(1, 1.toByte)
+    checkCast(1, 1.toShort)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:945]: are these tests non-ansi specific?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala [hunk 17]
```diff
@@ -930,150 +733,254 @@ abstract class CastSuiteBase extends SparkFunSuite with ExpressionEvalHelper {
...
+    checkCast(123L, "123")
+
+    checkEvaluation(cast(123L, DecimalType.USER_DEFAULT), Decimal(123))
+    checkEvaluation(cast(123L, DecimalType(3, 0)), Decimal(123))
+    checkEvaluation(cast(123L, DecimalType(3, 1)), null)
+
+    checkEvaluation(cast(123L, DecimalType(2, 0)), null)
+  }
+
+  test("cast from int 2") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:977]: `cast from int #2`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 15 above)
Comment (gengliangwang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala:1514]: Essentially, `withSQLConf` also uses `SQLConf`. There is no much difference here.
This is a common issue for test suites of the Catalyst module.  We should figure out a proper way for setting SQL conf without impacts on other concurrent tests later.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala (hunk 16 above)
Comment (gengliangwang) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CastSuite.scala]: Good catch. I have updated the code.

**Review by cloud-fan — APPROVED** [review]

**Comment by gengliangwang** [comment]
retest this please.

**Comment by maropu** [comment]
Thanks! Merged to master.

**Comment by dongjoon-hyun** [comment]
Thank you, @gengliangwang and all!

**Comment by revans2** [comment]
I just noticed that the error messages when trying to cast between timestamps and numeric types is wrong now in ANSI mode.

```
org.apache.spark.sql.AnalysisException: cannot resolve 'CAST(`c0` AS TIMESTAMP)' due to data type mismatch: cannot cast int to timestamp,you can enable the casting by setting spark.sql.legacy.allowCastNumericToTimestamp to true,but we strongly recommend using function TIMESTAMP_SECONDS/TIMESTAMP_MILLIS/TIMESTAMP_MICROS instead.;;
```

`spark.sql.legacy.allowCastNumericToTimestamp` has no impact on ANSI casts after this change.  It looks like a fairly simple fix.

https://github.com/apache/spark/blob/3695e997d5d436be086235505bbb030c87ae8eef/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L276

**Comment by gengliangwang** [comment]
@revans2 Thanks for reporting the problem. Actually, I am working on another PR to give suggestions for the disallowed ANSI cast. It will be sent out today and I will ping you in the PR.

**Comment by revans2** [comment]
@gengliangwang sounds good.  Thanks
